server/src/main/java/org/elasticsearch/index/query/MatchPhraseQueryBuilder.java

```diff
@@ -163,8 +163,9 @@ protected QueryBuilder doIndexMetadataRewrite(QueryRewriteContext context) throw
         // and possibly shortcut
         // If we're using a keyword analyzer then we can rewrite this to a TermQueryBuilder
         // and possibly shortcut
+        // If no analyzer is set then we can rewrite this to a TermQueryBuilder and possibly shortcut
         NamedAnalyzer configuredAnalyzer = configuredAnalyzer(context);
-        if (configuredAnalyzer != null && configuredAnalyzer.analyzer() instanceof KeywordAnalyzer) {
+        if (configuredAnalyzer == null || configuredAnalyzer.analyzer() instanceof KeywordAnalyzer) {
```

Comment:
you are using the analyzer check as a proxy to detect whether the field is unmapped? Should we be more explicit and check that the field is unmapped?